### ISSUE TITLE
Serokell: remove BuiltinAndAxiomSimplifier

### ISSUE THREAD

**Issue by ana-pantilie** [runtimeverification/haskell-backend#3111]
### Progress until 07/10/2022
- [x] Specialize BuiltinAndAxiomSimplifier to a concrete monad
- [x] Call builtin evaluator functions statically
- [ ] Remove BuiltinAndAxiomSimplifier by doing https://github.com/runtimeverification/haskell-backend/issues/3111#issuecomment-1210727293

**Comment by int-index** [comment]
## Background

`BuiltinAndAxiomSimplifier` is a data type defined in `Kore.Simplify.Simplify` as follows:

```haskell
newtype BuiltinAndAxiomSimplifier =
    BuiltinAndAxiomSimplifier
    { runBuiltinAndAxiomSimplifier ::
        forall simplifier.
        MonadSimplify simplifier =>
        TermLike RewritingVariableName ->
        SideCondition RewritingVariableName ->
        simplifier (AttemptedAxiom RewritingVariableName)
    }
```

Those simplifiers are stored in a map:

```haskell
type BuiltinAndAxiomSimplifierMap =
    Map.Map AxiomIdentifier BuiltinAndAxiomSimplifier
```

And there's a function to look up an appropriate simplifier for a termlike:

```haskell
lookupAxiomSimplifier ::
    MonadSimplify simplifier =>
    TermLike RewritingVariableName ->
    MaybeT simplifier BuiltinAndAxiomSimplifier
```

This machinery is then used in `maybeEvaluatePattern` (in `Kore.Rewrite.Function.Evaluator`):

```haskell
maybeEvaluatePattern childrenCondition termLike defaultValue sideCondition =
        do
            BuiltinAndAxiomSimplifier evaluator <- lookupAxiomSimplifier termLike
            lift $ do
                merged <- do
                    result <- evaluator termLike sideCondition
                    ...
```

This is important because our profiling data indicates that a lot of time is spent in this code path. More specifically, `evaluateApplication` calls `maybeEvaluatePattern`, which in turn calls `lookupAxiomSimplifier`, which finds a simplifier in the `simplifierAxioms` part of the `Env` (initialized in `mkSimplifierEnv`), and running those simplifiers is where a lot of time is spent.

## Problem

There are two issues with this approach that stand in the way of compiler optimizations. There are two important optimizations that GHC applies: specialization and inlining. The way things are currently set up, both of those are impossible.

### Specialization

`BuiltinAndAxiomSimplifier` defines a computation polymorphic in its monadic context and wraps it in a newtype. The compiler is not going to specialize this: the newtype fixes the runtime representation of the computation, and you are certain to end up with dictionary-passing for the `MonadSimplify` constraint.

The first step to fixing that is to define `BuiltinAndAxiomSimplifier` as follows:

```haskell
newtype BuiltinAndAxiomSimplifier =
    BuiltinAndAxiomSimplifier
    { runBuiltinAndAxiomSimplifier ::
        TermLike RewritingVariableName ->
        SideCondition RewritingVariableName ->
        Simplifier (AttemptedAxiom RewritingVariableName)
    }
```

Now we have a concrete monad — `Simplifier` — instead of an abstract `MonadSimplify simplifier => ...`. With this change, we can start specializing the actual computations that are wrapped in this newtype. Let's take a closer look at the way `simplifierAxioms` are constructed. We start in `mkSimplifierEnv` where we call `mkEvaluatorRegistry`, where we call `mkEvaluator`, where we call `simplificationEvaluation`, where we call `attemptEquation`. (I am talking about this specific code path because `attemptEquation` is also very prominent in the performance profile). The type of `attemptEquation` is:

```haskell
attemptEquation ::
    forall simplifier.
    HasCallStack =>
    MonadSimplify simplifier =>
    SideCondition RewritingVariableName ->
    TermLike RewritingVariableName ->
    Equation RewritingVariableName ->
    simplifier (AttemptEquationResult RewritingVariableName)
```

Note the abstract `simplifier` — we want a concrete `Simplifier` instead to help GHC optimize the code. Previously this wasn't an option because we needed a polymorphic `attemptEquation` to put into `BuiltinAndAxiomSimplifier`. Now we can specialize it (either by hand or with a pragma) so that our `simplifierAxioms` would contain an optimized, monomorphic version of this subroutine.

## Inlining

Another issue with `BuiltinAndAxiomSimplifierMap` is that it is, well, a map. We're looking up the simplifiers in this map at runtime (with `lookupAxiomSimplifie`), so GHC would never inline any of them, and that might block further optimizations. However, this is not necessary, as many of those simplification routines are known statically. Consider `koreEvaluators` in `Kore.Builtin`:

```haskell
koreEvaluators ::
    -- | Module under which evaluation takes place
    VerifiedModuleSyntax StepperAttributes ->
    Map AxiomIdentifier BuiltinAndAxiomSimplifier
koreEvaluators = evaluators builtins
  where
    builtins :: Map Text BuiltinAndAxiomSimplifier
    builtins =
        Map.unions
            [ Bool.builtinFunctions
            , Int.builtinFunctions
            , KEqual.builtinFunctions
            , List.builtinFunctions
            , Map.builtinFunctions
            , Set.builtinFunctions
            , String.builtinFunctions
            , Krypto.builtinFunctions
            , InternalBytes.builtinFunctions
            ]
```

All evaluators for builtins are known statically. Instead of building a `Map` with them, consider defining a single function like this:

```haskell
simplifyBuiltins ::
        TermLike RewritingVariableName ->
        SideCondition RewritingVariableName ->
        MaybeT Simplifier (AttemptedAxiom RewritingVariableName)
```

(I'm not sure if this is the right type signature, but it should convey the general idea).

And instead of taking a `Map.unions` of submaps, call functions directly:

```haskell
simplifyBuiltins' termLike =
  Bool.simplifyBuiltins termLike <|>
  Int.simplifyBuiltins termLike <|>
  KEqual.simplifyBuiltins termLike <|>
  ...
```

(Once again, the code would probably look differently; I'm just trying to contrast `Map` lookups with direct function calls).

# Conclusion

To sum it up, try to do the following:

1. Specialize `BuiltinAndAxiomSimplifier` to a concrete monad (preliminary experiments in #3153 indicate that it is indeed feasible).
2. Avoid `BuiltinAndAxiomSimplifierMap` when the simplification subroutines are known statically.

**Comment by goodlyrottenapple** [comment]
## A few more notes on the inlining of `BuiltinAndAxiomSimplifierMap`

The bulk of initialization of  `BuiltinAndAxiomSimplifierMap` seems to happen here:

https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Simplify/Data.hs#L260-L279

### Removing `builtinEvaluators`

We want to remove the `builtinEvaluators` map from the code below. This map is constructed via `Builtin.koreEvaluators` which takes in a `verifiedModule'` and constructs a mapping between [`AxiomIdentifier.Application Id`s](https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Rewrite/Axiom/Identifier.hs#L66-L68) of hooked symbol sentences in the module and the evaluators for the given hooked symbol (see the [evaluator](https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Builtin.hs#L138) function). Note that we lookup the hooked symbol string and match it to the builtins map here:

https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Builtin.hs#L169-L173

(`StepperAttributes` is just synonym for `Symbol`)


Then, during simplification of a `TermLike` inside the [`lookupAxiomSimplifier`](https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Simplify/Simplify.hs#L441), we call `matchAxiomIdentifier`

https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Rewrite/Axiom/Identifier.hs#L103-L106

which pulls out the symbol `Id` out of the given `TermLike` and looks it up in the map. When this `Id` points to a hooked symbol, instead of doing this round trip, we could probably match directly on the `Hook` inside the `Symbol` that `TermLike` carries.

### Potential issues

It is not clear to me if we ever actually apply `Axiom.EvaluationStrategy.simplifierWithFallback` when combining the builtin and user evaluator maps in the `initialize` function above. It would seem unlikely that the hooked symbols would have both a builtin evaluator and a user defined one? If this can indeed be the case, then  [`lookupAxiomSimplifier`](https://github.com/runtimeverification/haskell-backend/blob/eb60e9b80faab374c648aaf71285186ddc3b9c77/kore/src/Kore/Simplify/Simplify.hs#L441) would have to take this into consideration and inline the application of this strategy within its body.

**Comment by ana-pantilie** [comment]
> It would seem unlikely that the hooked symbols would have both a builtin evaluator and a user defined one?

They can definitely have user defined evaluators. We have many simplification rules for arithmetic operations (`+Int` etc.) or for map/set functions. It's actually a widespread practice to add simplification rules for when those hooked symbols are applied to symbolic terms.

**Comment by ana-pantilie** [comment]
All function evaluators are known statically. Specifically for function definitions provided by the user, we don't need to store a map of Id -> FuncEval because the actual procedure which applies function is the same, more exactly it's: https://github.com/runtimeverification/haskell-backend/blob/ac0e3c0c403bbca4d812d921f8783f4dfcdbcb29/kore/src/Kore/Equation/Application.hs#L109. We should be storing a map of Id -> Equation, because the only thing which changes between user defined functions is the equation.

**Comment by ana-pantilie** [comment]
We need the server to be set up in order to do accurate performance testing for this one. Blocked until that is done.

**Comment by ana-pantilie** [comment]
Our priorities have changed, and the code maintainability we would have gotten from continuing with this issue is no longer as valuable as actual performance improvements. We can consider it done.
